### PR TITLE
Better controll of frames per second

### LINKED ISSUE #11: Accurate fps option
Apply spark-video on the elephants_dream video and we will get 6 frames per second actually.

### PR DESCRIPTION
**Pull request by da-tubi**
+ Feature: New Options `frameStepSize` and `frameStepOffset`
+ Refactor: Minor code refactoring
+ Fix: Load columns specified in readDataSchema

closes #11 

Assuming that we stick to the fixed frameStepSize way to extract frames, the fps option might not be accurate. This pull request proposed a relatively good way to specify frameStepSize and frameStepOffset, if it does not meet our need, we could specify frameStepSize and frameStepOffset manually.

### CODE REVIEW

File: video/src/ai/eto/rikai/sql/spark/datasources/VideoOptions.scala [hunk 1]
```diff
@@ -1,21 +1,65 @@
 package ai.eto.rikai.sql.spark.datasources
```
Comment (Renkai) [inline video/src/ai/eto/rikai/sql/spark/datasources/VideoOptions.scala:1]: My IDE say it `Package name 'ai.eto.rikai.sql.spark.datasources' does not correspond to the file path '' `, maybe it's legal in mill, but still confusing readers, how about changing the file position to make it align with package name.

File: video/src/ai/eto/rikai/sql/spark/datasources/VideoOptions.scala [hunk 2]
```diff
@@ -1,21 +1,65 @@
...
 class VideoOptions(@transient val parameters: CaseInsensitiveMap[String])
     extends Serializable {
   def this(parameters: Map[String, String]) = {
     this(CaseInsensitiveMap(parameters))
   }
 
+  // Number of frames per second
   val fps = parameters.get("fps").map(_.toInt).getOrElse(1)
+  // Frame step size and the frame_id to start
+  val frameStepSize = parameters.get("frameStepSize").map(_.toInt).getOrElse(0)
```
Comment (Renkai) [inline video/src/ai/eto/rikai/sql/spark/datasources/VideoOptions.scala:25]: How about using
```
  val frameStepSize : Option[Int] = parameters.get("frameStepSize").map(_.toInt)
  val frameStartId : Option[Int] = parameters.get("frameStartId").map(_.toInt)
```
`0` might be a valid value for some parameters, so we need to distinguish between "user have not filled this parameter" and "user fill this parameter with 0", is it a valid value for `frameStartId`?

Comment (da-tubi) [reply video/src/ai/eto/rikai/sql/spark/datasources/VideoOptions.scala:16]: in this case, 0 is a valid parameter

Comment (da-tubi) [reply video/src/ai/eto/rikai/sql/spark/datasources/VideoOptions.scala:25]: parameters.get("xxx") is already an Option. 0 is a default value here.

Comment (da-tubi) [reply video/src/ai/eto/rikai/sql/spark/datasources/VideoOptions.scala:25]: Let me add more inline comments

File: video/src/ai/eto/rikai/sql/spark/datasources/VideoPartitionReaderFactory.scala [hunk 3]
```diff
@@ -98,40 +63,64 @@ case class VideoPartitionReaderFactory(
...
+
+    val videoReader = new PartitionReader[InternalRow] {
       override def next(): Boolean = {
-        val targetFrameId = frame_id + numOfFps
-        while (frame_id < targetFrameId - 1) {
-          frame_id = frame_id + 1
+        val targetFrameId = frameId + frameStep
+        while (frameId < targetFrameId - 1) {
+          frameId = frameId + 1
           grabber.grabImage()
```
Comment (Renkai) [inline video/src/ai/eto/rikai/sql/spark/datasources/VideoPartitionReaderFactory.scala:99]: I saw `grabImage()` here equals to `grabFrame(false, true, true, false, false)`, will change some of this parameter make it more efficient?

File: video/src/ai/eto/rikai/sql/spark/datasources/VideoPartitionReaderFactory.scala [hunk 4]
```diff
@@ -98,40 +63,64 @@ case class VideoPartitionReaderFactory(
...
-        row.update(2, frame.timestamp)
-        val javaImage = converter.convert(frame)
-        val bos = new ByteArrayOutputStream()
-        ImageIO.write(javaImage, "png", bos)
-        row.update(3, bos.toByteArray)
+        val size = readDataSchema.size
+        val row = new GenericInternalRow(size)
+        (0 until size).foreach { index =>
+          readDataSchema(index).name match {
+            case "video_uri" =>
```
Comment (Renkai) [inline video/src/ai/eto/rikai/sql/spark/datasources/VideoPartitionReaderFactory.scala]: Is it good to make the field names an enum in some common package?

File: video/src/ai/eto/rikai/sql/spark/datasources/VideoPartitionReaderFactory.scala (hunk 3 above)
Comment (da-tubi) [reply video/src/ai/eto/rikai/sql/spark/datasources/VideoPartitionReaderFactory.scala:99]: I don't know yet. Let us try in another performance tuning pull request.

**Comment by Renkai** [comment]
The rest LGTM

File: video/src/ai/eto/rikai/sql/spark/datasources/VideoOptions.scala [hunk 5]
```diff
@@ -1,21 +1,65 @@
...
   val imageHeight = parameters.get("imageHeight").map(_.toInt).getOrElse(0)
 
   /** A list of scaler flags from
     * http://www.ffmpeg.org/ffmpeg-scaler.html#toc-Scaler-Options
     */
   val scalerFlag = parameters.get("scalerFlag").getOrElse("bilinear")
+
+  def getFrameStep(grabber: FrameGrabber): (Int, Int) = {
+    if (frameStepSize > 0) {
+      val startId = if (frameStartId == 0) frameStepSize else frameStartId
```
Comment (Renkai) [inline video/src/ai/eto/rikai/sql/spark/datasources/VideoOptions.scala]: Why user will not need a frame start id `0`?

Comment (da-tubi) [reply video/src/ai/eto/rikai/sql/spark/datasources/VideoOptions.scala]: Because frame_id starts from 1

Comment (da-tubi) [reply video/src/ai/eto/rikai/sql/spark/datasources/VideoOptions.scala]: It should be documented.

**Review by Renkai — APPROVED** [review]

**Pull request merged**